File: 6-zigzag-conversion/6-zigzag-conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        return solve(s, numRows);
    }
private:
    // Self
    string solve(string& s, int numRows) {
        if(numRows <= 1) return s;
        vector<vector<char>> res(numRows, vector<char>(1001));
        
        int dir = 1;
        int i = 0, j = 0;
        
        for(auto ch: s) {
            res[i][j] = ch;
            
            i += dir;
            
            // Changing the direction to upwards/downwards
            if(i == numRows){
                i -= 2;
                dir *= -1;
            }
            else if(i == -1) {
                i += 2;
                dir *= -1;
            }
            // if(i < 0) i = 0;
            // If direction is upward then change column
            if(dir == -1)
                ++j;
        }
        // for(auto x: res) {
        //     for(auto y: x)
        //         cout<<y<<" ";
        //     cout<<endl;
        // } 
        string ans = "";
        for(auto row: res) {
            for(auto ch: row) {
                if(isalpha(ch) || ch == ',' || ch == '.')
                    ans += ch;
            }
        }
        return ans;
    }
};

File: 6-zigzag-conversion/6-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows <= 1:
            return s
        
        res = [['-'] * 1001 for _ in range(numRows)]
        dr = 1
        i, j = 0, 0
        
        for ch in s:
            res[i][j] = ch
            
            i += dr
            
            if i == numRows:
                i -= 2
                dr *= -1
            
            elif i == -1:
                i += 2
                dr *= -1
            
            if dr == -1:
                j += 1
                
        ans = []
        for row in res:
            for ch in row:
                if ch != '-':
                    ans.append(ch)
        return ''.join(ans)